### PR TITLE
[quant][graphmode][fx] Support dynamic quantization without calibration

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43984 [quant][graphmode][fx][api] Call fuse in prepare
* #43983 [quant][graphmode][fx] Support inplace option
* **#43952 [quant][graphmode][fx] Support dynamic quantization without calibration**

Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

### CODE REVIEW

File: torch/quantization/fx/quantize.py [hunk 1]
```diff
@@ -266,10 +266,31 @@ def prepare(self, model, qconfig_dict, inplace=False):
     def prepare_dynamic(self, model, qconfig_dict, inplace=False):
         return self._prepare(model, qconfig_dict, inplace, is_dynamic_quant=True)
 
+    # Extract the subgraph that produces the weight for dynamically quantized node
+    # and run the subgraph to observe the weight
+    def _run_weight_observers(self, observed):
+        for node in observed.graph.nodes:
+            if node.op == 'call_function' and node.target in WEIGHT_INDEX_DICT:
```
Comment (supriyar) [inline torch/quantization/fx/quantize.py:273]: How does this work for dynamic quantizable modules?

File: torch/quantization/fx/quantize.py [hunk 2]
```diff
@@ -266,10 +266,31 @@ def prepare(self, model, qconfig_dict, inplace=False):
...
+                for i, node_arg in enumerate(node.args):
+                    if i in WEIGHT_INDEX_DICT[node.target]:
```
Comment (supriyar) [inline torch/quantization/fx/quantize.py:275]: Should there be a separate dict for dynamic ops?

File: torch/quantization/fx/quantize.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:273]: for dynamic quantizable modules, the weight observer is run during conversion: e.g. https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/dynamic/modules/linear.py#L96

File: torch/quantization/fx/quantize.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:275]: not sure, isn't the weight index the same for static and dynamic?

File: torch/quantization/fx/quantize.py (hunk 1 above)
Comment (supriyar) [reply torch/quantization/fx/quantize.py:273]: Can we add a comment here which mentions this?

File: torch/quantization/fx/quantize.py (hunk 2 above)
Comment (supriyar) [reply torch/quantization/fx/quantize.py:275]: I think we should only run weight observers for dynamic quantizable ops, the dict has both static and dynamic ops.

File: torch/quantization/fx/quantize.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:273]: sure

File: torch/quantization/fx/quantize.py (hunk 2 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:275]: this pass is only called in dynamic quantization. WEIGHT_INDEX_DICT is containing floating point ops, we haven't produced static/dynamic quantized ops at this point.

**Review by supriyar — APPROVED** [review]

**Pull request closed without merging**